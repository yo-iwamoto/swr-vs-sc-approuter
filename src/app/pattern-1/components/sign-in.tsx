"use client";

import { useNotification } from "@/app/components/notification-bar-area";
import type { FormEvent } from "react";
import { Base, Button, FormControl, Input } from "smarthr-ui";
import { mutate } from "swr";
import { useSignInMutation } from "@/app/pattern-1/mutations/use-sign-in-mutation";

export function SignIn() {
  const { notify } = useNotification();
  const signInMutation = useSignInMutation();

  const onSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const username = new FormData(e.currentTarget).get("username");
    if (typeof username !== "string") return;

    const res = await signInMutation.trigger({ username });
    await mutate(["me"]);
    const message =
      res.type === "signin"
        ? "サインインしました"
        : "ユーザーを新しく作成し、サインインしました";
    notify({ type: "success", message });
  };

  return (
    <Base className="p-8 max-w-md mx-auto my-20">
      <form onSubmit={onSubmit} className="flex flex-col items-center gap-2">
        <h1 className="text-lg font-bold">サインイン画面</h1>

        <FormControl title="ユーザー名" className="w-full">
          <Input
            name="username"
            required
            maxLength={100}
            className="w-full py-2"
            data-1p-ignore
          />
        </FormControl>

        <Button
          type="submit"
          variant="primary"
          className="w-full"
          loading={signInMutation.isMutating}
        >
          <span>サインイン</span>
        </Button>
      </form>
    </Base>
  );
}
