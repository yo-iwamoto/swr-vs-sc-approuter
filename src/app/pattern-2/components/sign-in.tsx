import { Base, Button, FormControl, Input } from "@/components/client-ui";
import { envVars } from "@/config/env";
import { prisma } from "@/lib/prisma";
import * as jwt from "hono/jwt";
import { cookies } from "next/headers";

export function SignIn() {
  const signInAction = async (data: FormData) => {
    "use server";

    const username = data.get("username");
    if (typeof username !== "string") return;

    const user = await prisma.user.findUnique({ where: { username } });
    if (user === null) {
      const newUser = await prisma.user.create({ data: { username } });
      const token = await jwt.sign(
        { userId: newUser.id },
        envVars.JWT_SECRET,
        "HS256",
      );
      cookies().set("token", token);
      return;
    }

    const token = await jwt.sign({ userId: user.id }, envVars.JWT_SECRET);
    cookies().set("token", token);
  };

  return (
    <Base className="p-8 max-w-md mx-auto my-20">
      <form action={signInAction} className="flex flex-col items-center gap-5">
        <h1 className="text-lg font-bold">サインイン画面</h1>

        <p className="text-sm text-center">
          任意のユーザー名でサインインできます。
          <br />
          存在しないユーザー名の場合、作成してサインインします。
        </p>

        <FormControl title="ユーザー名" className="w-full">
          <Input
            name="username"
            required
            maxLength={100}
            className="w-full py-2"
            data-1p-ignore
          />
        </FormControl>

        <Button type="submit" variant="primary" className="w-full">
          <span>サインイン</span>
        </Button>
      </form>
    </Base>
  );
}
