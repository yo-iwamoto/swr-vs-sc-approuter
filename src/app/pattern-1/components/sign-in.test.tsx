import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SignIn } from "./sign-in";

const triggerMock = vi.fn(async () => ({ type: "signin" }));
vi.mock("@/app/pattern-1/mutations/use-sign-in-mutation", () => ({
  useSignInMutation: vi.fn(() => ({
    trigger: triggerMock,
  })),
}));

const mutateMock = vi.hoisted(() => vi.fn());
vi.mock("swr", () => ({ mutate: mutateMock }));

const notifyMock = vi.hoisted(() => vi.fn());
vi.mock("@/app/components/notification-bar-area", () => ({
  useNotification: vi.fn(() => ({
    notify: notifyMock,
  })),
}));

describe("SignIn", () => {
  it("サインインボタンを押すと、サインインリクエストが行われること", async () => {
    render(<SignIn />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ユーザー名" }),
      "user",
    );
    await userEvent.click(screen.getByRole("button", { name: "サインイン" }));

    expect(triggerMock).toHaveBeenLastCalledWith({ username: "user" });
  });

  it('サインイン後、"me" が mutate されること', async () => {
    render(<SignIn />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ユーザー名" }),
      "user",
    );
    await userEvent.click(screen.getByRole("button", { name: "サインイン" }));

    expect(mutateMock).toHaveBeenLastCalledWith(["me"]);
  });

  it("サインイン後、種別がサインインであれば、サインインした旨通知が表示されること", async () => {
    render(<SignIn />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ユーザー名" }),
      "user",
    );
    await userEvent.click(screen.getByRole("button", { name: "サインイン" }));

    expect(notifyMock).toHaveBeenLastCalledWith({
      type: "success",
      message: "サインインしました",
    });
  });

  it("サインイン後、種別がサインアップであれば、ユーザーを作成した旨通知が表示されること", async () => {
    triggerMock.mockResolvedValue({ type: "signup" });

    render(<SignIn />);

    await userEvent.type(
      screen.getByRole("textbox", { name: "ユーザー名" }),
      "user",
    );
    await userEvent.click(screen.getByRole("button", { name: "サインイン" }));

    expect(notifyMock).toHaveBeenLastCalledWith({
      type: "success",
      message: "ユーザーを新しく作成し、サインインしました",
    });
  });
});
