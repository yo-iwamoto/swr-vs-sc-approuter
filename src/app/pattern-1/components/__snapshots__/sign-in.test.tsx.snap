// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignIn > snapshot 1`] = `
<DocumentFragment>
  <div
    class="smarthr-ui-Base shr-bg-white forced-colors:shr-border-shorthand contrast-more:shr-border-high-contrast shr-rounded-l shr-shadow-layer-1 p-8 max-w-md mx-auto my-20"
  >
    <form
      class="flex flex-col items-center gap-5"
    >
      <h1
        class="text-lg font-bold"
      >
        サインイン画面
      </h1>
      <p
        class="text-sm text-center"
      >
        任意のユーザー名でサインインできます。
        <br />
        存在しないユーザー名の場合、作成してサインインします。
      </p>
      <div
        class="shr-flex-col shr-justify-start [&_>_*]:shr-my-0 shr-flex shr-space-y-0.5 sc-ilxebA fSuyNt w-full smarthr-ui-FormGroup"
      >
        <label
          class="shr-flex-wrap [&:empty]:shr-gap-0 shr-flex shr-gap-y-0.5 shr-gap-x-0.5 shr-items-center sc-jIBlqr hndDbA smarthr-ui-FormGroup-label"
          for=":r6:"
          id=":r7:"
        >
          <span
            class="shr-text-base shr-font-bold"
          >
            ユーザー名
          </span>
        </label>
        <div
          class="sc-gYrqIg keLuCv"
        >
          <span
            class="smarthr-ui-Input shr-border-shorthand shr-box-border shr-inline-flex shr-cursor-text shr-items-center shr-gap-0.5 shr-rounded-m shr-bg-white shr-px-0.5 contrast-more:shr-border-high-contrast focus-within:shr-focus-indicator w-full py-2"
            role="presentation"
          >
            <input
              class="smarthr-ui-Input-input shr-inline-block shr-w-full shr-grow shr-border-none shr-bg-transparent shr-py-0.75 shr-text-base shr-leading-none shr-text-black shr-outline-none shr-outline-0 placeholder:shr-text-grey disabled:shr-text-disabled disabled:shr-opacity-100 shr-h-[theme(fontSize.base)]"
              data-1p-ignore="true"
              id=":r6:"
              maxlength="100"
              name="username"
              required=""
            />
          </span>
        </div>
      </div>
      <button
        class="disabled:shr-cursor-not-allowed disabled:shr-bg-clip-padding [&_.smarthr-ui-Icon]:forced-colors:disabled:shr-fill-[GrayText] shr-box-border shr-cursor-pointer shr-inline-flex shr-justify-center shr-items-center shr-gap-0.5 shr-text-center shr-whitespace-nowrap shr-rounded-m shr-border-shorthand shr-font-inherit shr-font-bold shr-leading-none focus-visible:shr-focus-indicator contrast-more:shr-border-high-contrast [&_.smarthr-ui-Icon]:shr-block [&_svg]:shr-block shr-text-base shr-px-1 shr-py-0.75 shr-border-main shr-bg-main shr-text-white focus-visible:shr-border-main-darken focus-visible:shr-bg-main-darken hover:shr-border-main-darken hover:shr-bg-main-darken disabled:shr-border-main/50 disabled:shr-bg-main/50 disabled:shr-text-white/50 smarthr-ui-Button w-full"
        type="submit"
      >
        <span
          class="shr-min-w-0"
        >
          <span>
            サインイン
          </span>
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
